Vsd meet and assume

jezhiggins opened the pull request:
Two related strands of work in this PR. 

* Extending `abstract_environmentt::assume` to cover all value representations. Up until now, `assume` was only correct for constants. It _may_ have evaluated a given expression correctly for intervals and value-sets, but not reliably. These changes all assume expressions to be properly evaluated for all value representation, and in the face of different representations within the same expression.

* Filling out the implementation of `abstract_objectt::meet` for value representations. Where `merge` is the union of two values, `meet` is the intersection. `meet` is used when evaluating an assumed equality expression. 

+2,331/-384 looks alarming, but the major bulk of that is test cases exercising various combinations of constants, intervals, value-sets, and operators in assume and in meet.

jezhiggins committed:
abstract_object::merge can take its parameters by const ref

jezhiggins committed:
Characterising cross-representation equals and not equals expressions

jezhiggins committed:
Return the proper interval when an object_set contains booleans

jezhiggins committed:
Split out do_assume from abstract_environment::assume

Seperates the assume evaluation from the action arising.

jezhiggins committed:
Rewrite comparison tests in terms of environment::do_assume

jezhiggins committed:
Pass const references into abstract_objectt::meet

No need to copy the arguments.

jezhiggins committed:
abstract_objectt::meet - if this is top, return the other object

The meet of TOP is the other value,
  meet(TOP, 1) is 1
  meet(1, TOP) is 1
  meet(TOP, TOP) is TOP

jezhiggins committed:
Implement meet for constant_abstract_value

jezhiggins committed:
Extend interval_abstract_valuet::meet for all value representations

jezhiggins committed:
Initial assume_eq

This is the first step on splitting out assume so we can given sensible
answers for all value representations, not just constants.

jezhiggins committed:
value_set_abstract_objectt::meet implementation

Relatively simple - converts both objects to an interval, and computes
the intersection. Returns a value set containing that intersection -
either a constant, or an interval.

jezhiggins committed:
unit tests for assume, equality comparison

jezhiggins committed:
Refactor assume unit tests so we can more easily add further tests

jezhiggins committed:
First pass at evaluating != in assume.

jezhiggins committed:
Extend assume to handle negated expressions.

Consider cases when one or other operand is not a value, or is unknown.

Implement meet for write_location_context.

jezhiggins committed:
Extend assume to cover <=, less than or equals

jezhiggins committed:
Extend assume to cover <, less than

jezhiggins committed:
Replace if ladder with table lookup in do_assume

jezhiggins committed:
Pulled out eval_operands_as_values from assume_le/assume_lt

jezhiggins committed:
Extend assume to cover >=, greater than or equal

jezhiggins committed:
Extended assume to cover >, greater than

jezhiggins committed:
Extended assume to cover and expressions

jezhiggins committed:
Extended assume to cover or expressions

jezhiggins commented:
MacOS build failures look to be a meta-problem getting the tool dependencies 

```
==> Downloading https://homebrew.bintray.com/bottles/ninja-1.10.2.catalina.bottle.tar.gz
curl: (22) The requested URL returned error: 403 Forbidden
Error: Failed to download resource "ninja"
Download failed: https://homebrew.bintray.com/bottles/ninja-1.10.2.catalina.bottle.tar.gz
Error: Process completed with exit code 1.
```

martin-cs commented:
Paging @NlightNFotis and @thomasspriggs for the Mac OS CI issues.

NlightNFotis commented:
I took a quick look at it - in the beginning I thought that they upgraded the version so it's hitting the wrong URL because of outdated database issues (kind of like needing an apt-get update, if you want).

But it looks like it's still the same version of the formula (https://formulae.brew.sh/formula/ninja), and it's 403 if you hit the URL, which seems to be sitting on their servers, so I'm assuming this must be a temporary issue they are having.

Could you try it again (just restart the jobs from the github interface) later, and let me know if it doesn't work? I'll take a deeper look at it then.

tautschnig commented:
> MacOS build failures look to be a meta-problem getting the tool dependencies
> 
> ```
> ==> Downloading https://homebrew.bintray.com/bottles/ninja-1.10.2.catalina.bottle.tar.gz
> curl: (22) The requested URL returned error: 403 Forbidden
> Error: Failed to download resource "ninja"
> Download failed: https://homebrew.bintray.com/bottles/ninja-1.10.2.catalina.bottle.tar.gz
> Error: Process completed with exit code 1.
> ```

For the record: https://github.com/actions/virtual-environments/issues/3165 - presumably the current run will have updated runners already.

jezhiggins committed:
Test showing loop terminates using constants, intervals, or value-sets

jezhiggins committed:
Correct evaluation of assume in the face of not_exprs

For intervals and value-sets it's possible for, say, (a > b) and !(a >
b) to both be true. When evaluationing !(expr) we can't simply flip the
result. Instead, where we can, we need to rewrite !(a > b) as (a <= b)
and evaluate that instead.

martin-cs commented on regression/goto-analyzer/branching-ge/test-always-intervals.desc:
```diff
@@ -0,0 +1,8 @@
+^EXIT=0$
+^SIGNAL=0$
+^main::1::i .* \[5, 5\] @ \[17\]
+^main::1::p .* \[0, A\] @ \[3, 8, 14\]
```
One day, I will have to fix the printing in hex...

jezhiggins replied:
This is a pretty tiny change, but the main reason I haven't is the donkey work of updating a bunch of regression tests :)

martin-cs replied:
Yep!  That was what was discouraging me too.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -196,21 +225,18 @@ bool abstract_environmentt::assume(const exprt &expr, const namespacet &ns)
+    if(assumption.is_false())
     {
       bool currently_bottom = is_bottom();
       make_bottom();
```
The comment block a few lines down can be removed.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -239,6 +265,29 @@ bool abstract_environmentt::assume(const exprt &expr, const namespacet &ns)
+                                      {ID_ge, assume_greater_than},
+                                      {ID_gt, assume_greater_than}};
+
+exprt abstract_environmentt::do_assume(const exprt &expr, const namespacet &ns)
```
Is this intended to only be used on expressions of boolean type?  If so, please can you add a PRECONDITION to document this.

jezhiggins replied:
No, it will operate on any expression. It's not a boolean expression, it falls back to evaluating it in exactly the same way as before.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +517,210 @@ std::vector<abstract_object_pointert> eval_operands(
+}
+
+static auto inverse_operations =
+  std::map<dstringt, dstringt>{{ID_equal, ID_notequal},
```
`irep_idt` rather than raw `dstring`.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +517,210 @@ std::vector<abstract_object_pointert> eval_operands(
+
+  auto result = env.do_assume(not_expr.op(), ns);
+  if(result.is_boolean())
+    result = result.is_true() ? exprt(false_exprt()) : true_exprt();
```
Why the `exprt` constructor on one side?

jezhiggins replied:
It's a cast rather than a constructor. 
The ternary operator wants both operands to have the same type, but `result.is_true() ? false_exprt() : true_exprt()` fails because `false_exprt` and `true_exprt` are different types (even though both can be converted to the `exprt`type `result` wants. By forcing the first operand to `exprt`, the second operand decays to `exprt` automatically and it all compiles and runs. 
I've replaced with a `static_cast`.

martin-cs replied:
Ahhhhhh.... yeahhhhh... ugh; the C++ type system and when subtyping relations are taken.  That is an annoying wrinkle. Thank you.

thomasspriggs replied:
Using `static_cast` isn't a particularly good solution either as it is an unchecked cast. `expr_checked_cast` would not currently work either. It would need its type signature tweaking slightly and `can_cast_expr<exprt>` to be defined. I suggest just using an `if` statement instead of the ternary operator here.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +517,210 @@ std::vector<abstract_object_pointert> eval_operands(
+  if(meet->is_bottom())
+    return true_exprt();
+
+  return false_exprt();
```
Is this correct?  If left is `{0, 1, 2}` and right is `{1, 2, 3}`, the meet is `{1,2}` which is not enough to conclude anything.  We can only return `false_exprt()` if both left and right are the same, single possible value.

jezhiggins replied:
I believe it is, but that doesn't mean it is. :slightly_smiling_face: 

For the equals case, if objects meet we consider that there exists some equality and so we return true. If the meet is empty, they're not equal and we return false. 
From our conversations, I understood the != case to be essentially equivalent. If the object do not meet, they are not equal and we return true. If there's some meet they are not not equal and we return false. 

Reworking this to return `false_exprt()` only when left and right are the same and `nil_exprt` otherwise breaks (or at least changes) existing behaviour in, for example, some of the simplification regression tests. It also causes segfaults in some of the regression tests I wrote around this work. 

The code, as written here, behaves better than the previous code in many cases and no worse than the previous code in every other case. It may not be absolutely correct, but I feel we can still merge this change, gain the benefits, and then revisit as work continues.

martin-cs replied:
There is an asymmetry which happens because the domains are representing /possible/ values for each variable rather than values they definitely can take.

So in the equal cases, we can take the meet of the two domains and get all of the values at which they could both have and be equal.  As you say, if it is empty then there are no ways they can be equal, so we return false and set the domain to bottom.  If there are things in the meet then it is possible they are equal and /we know that they must have a value from the meet/, so we can prune as you do.


The inequality case is a little bit harder and less useful.  If we take the meet of the values and it is empty then there was never any chance they were equal, so we return true because that constraint really is always true.  That much is symmetric.  However if the meet is non-empty (not bottom) then all is says is that it is /possible/ that they could be equal.  It doesn't mean they will be.  In the example I give, if the meet is `{1,2}` then it is possible that `left` is `1` and `right` is `2` or vica versa.  The only cases we need to remove are when they /have/ to be equal.  However that is really hard to represent if you have a set of alternatives for `left` and a separate one for `right` (this is where the non-relational part of the abstraction is an impediment, we don't have a way of representing the relationship between them).  To prune out possibilities we need to check when things are forced to be equal.  This only happens when one side only has one element (i.e. `to_constant` returns something other than `ID_nil`).  If the other side is also one element and the same one (meet is not empty) then they are necessarily equal, so we can return false as assuming they are not equal rules out all possibilities.  ( There is also a pruning case, where one side has one possibility and the other has multiple ones, in which case you can remove that value.  Easy-ish for value-set, possible for intervals if it is the boundary, otherwise not really possible. )

Whether this gets fixed as part of this PR or as a follow up ... either is fine.  I have a mild preference for fixing it now, esp. given the other changes I am requesting to this area.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +517,210 @@ std::vector<abstract_object_pointert> eval_operands(
+  return {as_value(left), as_value(right)};
+}
+
+exprt assume_less_than(
```
Is it worth a comment saying this can be ever-so slightly tightened by considering the cases of less-than and less-than-or-equal-to separately.

jezhiggins replied:
We do consider the less-than and less-then-or-equal separately (also greater-than and greater-than-or-equal) - when we rewrite the expression in simplified form we use the operator ID from the original expression. We don't change the comparison, we just simplify the operands.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +517,210 @@ std::vector<abstract_object_pointert> eval_operands(
+
+  auto left_upper = operands.left_interval().get_upper();
+  auto right_lower = operands.right_interval().get_lower();
+
```
Am I right in thinking that this will not resolve the following:

```
int x = nondet();

if (x < 10) {
  assert(x < 20);
}
```

because it does not prune the domain; it just identifies when a branch is infeasible (returning true / false)?

jezhiggins replied:
That's right, the assert will be UNKNOWN. We only prune the domain in the case of ==

martin-cs replied:
It would be really good if we could prune the domain on this case, it really makes a difference for precision and termination.  If there is no overlap then we return true or false (depending on if it is always greater than or never greater than).  If the two overlap (i.e. their meet is non-empty) then you can set the new bounds from the meet.

For example, if `x > y` then you can assign `x` to `[x.upper, y.upper]` and `y` to `[x.lower, y.lower]` (with the off-by-one for `>` vs. `>=`).  It might require me a little thought about how to do this in an abstraction agnostic way.

martin-cs reviewed (approved):
In principle happy but there are a few details to check.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -196,49 +225,51 @@ bool abstract_environmentt::assume(const exprt &expr, const namespacet &ns)
 }
 
+static auto assume_functions =
+  std::map<dstringt, assume_function>{{ID_not, assume_not},
```
s/dstring/irept/

`dstring` is the implementation but `irep_idt` is the "what irept.id stores" type.  At points there were other options apart from `dstring`.

jezhiggins replied:
I should have picked this up with the other dstring change :(

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -196,49 +225,51 @@ bool abstract_environmentt::assume(const exprt &expr, const namespacet &ns)
-  abstract_object_pointert res = eval(expr, ns);
-
-  exprt possibly_constant = res->to_constant();
+  auto simplified = simplify_expr(expr, ns);
```
I wonder if it makes sense to `ai_simplify` this instead?  That way anything that is directly constant should be substituted out and also used in the rewriting based simplification.  For example if we have `x + y + z == 0` and in the domain we know that `y ==0` and `z == 0`, `ai_simplify` should get us `x == 0`, while just `simplify_expr` on it's own won't.

jezhiggins replied:
`ai_simplify` isn't available here because that's a member function on `variable_sensitivity_domain`. Luckily, we call `assume` from `variable_sensitivity_domain`, so we can just call `assume` first :slightly_smiling_face: So I've done that.

martin-cs replied:
Great!

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -196,49 +225,51 @@ bool abstract_environmentt::assume(const exprt &expr, const namespacet &ns)
+  if(fn)
+    return fn(*this, expr, ns);
+
+  return eval(expr, ns)->to_constant();
```
This is a good fall-back option.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -196,49 +225,51 @@ bool abstract_environmentt::assume(const exprt &expr, const namespacet &ns)
+                                      {ID_ge, assume_greater_than},
+                                      {ID_gt, assume_greater_than}};
+
+exprt abstract_environmentt::do_assume(const exprt &expr, const namespacet &ns)
```
Can we document somewhere that returning an `exprt` with `ID_nil` is "give up on this assumption" and that `ID_false` sets the domain to bottom.  It is obvious when reading `assume` but I think it helps in understanding this code.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +499,211 @@ std::vector<abstract_object_pointert> eval_operands(
+  auto expr_id = expr.id();
+  auto inverse_operation = inverse_operations.find(expr_id);
+  if(inverse_operation == inverse_operations.end())
+    return nil_exprt();
```
There is a subtle precision loss here; if we have `if (A || B)`  then when we performing the transform for the `GOTO` instructions we will wind up assuming `not_exprt(or_exprt(A, B))` which will fail here rather than assuming `and_exprt(not_exprt(A), not_exprt(B))`.

You can either extend `invert_exprt` and get it to handle logical operations... or (which might be simpler) have an extra boolean flag argument for "we are assuming this to be ..." which starts as true.  `assume_not` will flip this (as will `assume_or` see below) and in the handlers for any atomic condition, you check the flag and if it is false, invert_exprt and recurse.

jezhiggins replied:
Having rewritten `assume_or` in terms of `assume_and`, I think this is taken care of. If we're assuming `not_exprt(or_exprt(A, B))` then that will be transformed to (more or less) `not_expr(not_expr(and_exprt(not_exprt(A), not_exprt(B))))`. It's a little long winded, but I think it will evaluate correctly.

martin-cs replied:
Sounds good; thanks.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +499,211 @@ std::vector<abstract_object_pointert> eval_operands(
+  for(auto const &operand : and_expr.operands())
+  {
+    auto result = env.do_assume(operand, ns);
+    if(result.is_nil() || result.is_false())
```
If `is_nil()` you can continue with the other operands.  It's only worth returning early if it is `is_false()`.

jezhiggins replied:
:+1:

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +499,211 @@ std::vector<abstract_object_pointert> eval_operands(
+  auto or_expr = to_or_expr(expr);
+  for(auto const &operand : or_expr.operands())
+  {
+    auto result = env.do_assume(operand, ns);
```
It is upsetting that this is not the dual of `assume_and` (and I fear I may not have communicated that very well) but this is not correct in the cases when `do_assume` can prune things (see comments below about pruning on ordering).  Consider the case of assuming `x == 1 || x == 2` as written this will (I think) result in the domain being bottom.

Instead I would have `assume_or(env, or_expr, ns, assuming_true)` call `assume_and(env, and_exprt(or_expr,operand()), ns, ! assume_true)`, i.e. you do the same thing as `assume_and` but inverting the each of the operands.

So rather than assuming `A || B`, you assume `!A && !B`, which means that if you prune the domain while handling `!A`, that is a good thing rather than conflicting with any of the pruning you do in `!B`.

jezhiggins replied:
I've rewritten `assume_or` in terms of `assume_and`.

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -468,3 +499,211 @@ std::vector<abstract_object_pointert> eval_operands(
+  return env.eval(reduced_le_expr, ns)->to_constant();
+}
+
+exprt assume_greater_than(
```
One of the symmetries that does work is `(a > b) == (b < a)` so if you want you can implement one of greater and less and simply recurse, flipping the arguments, to the other.

martin-cs reviewed (approved):
Sorry; I've done that awful reviewer thing of responding to your comments and then finding a bunch more things based on that.

If it is more efficient to talk these through, I am happy to do so.

jezhiggins commented:
> Sorry; I've done that awful reviewer thing of responding to your comments and then finding a bunch more things based on that.

That's ok. I'll work through these, and come back to you if there are any tricky bits :slightly_smiling_face:

jezhiggins committed:
Correct assume_and implementation

We can bail out early if an operand evals to false.
Otherwise we must evaluate al lthe operands, so that any pruning is
applied. That done, we return true iff all operands are true, otherwise
we return nil if anywhere nil.

jezhiggins committed:
Call ai_simplify before calling abstract_state.assume

jezhiggins committed:
Rewrite assume_or in terms of assume_and

jezhiggins committed:
Unit tests for assume value pruning

jezhiggins committed:
Domain pruning for < and <= assumptions.

Implemented for intervals and value-sets which are not top.

jezhiggins committed:
Reimplemented assume_greater_than in terms of assume_less_than

These operations are symmetric, so we can rewrite x > y as y < x, and
x >= y as y <= x.  As a consequence of rewriting, greater than
assumptions now also prune.

jezhiggins committed:
interval_abstract_value::output fix displaying upper limit when lower is -INF

jezhiggins committed:
Prune TOP values when assuming less than expressions

Remove duplication in data_dependency & write_location
merge and meet. Correct data_dependency_contextt::
has_been_modified to verify child object is also unchanged

jezhiggins commented:
@NlightNFotis @TGWDB I've got four of my regression tests that fail only on the VS with Make build. 

The failing tests all involve pruning a constant value TOP down to an interval where one bound is +/-MAX_INT. Curiously, when using intervals or value sets, the tests pass even though they will be pruned to the same interval. I have a suspicion, therefore, this may be related the typet used to construct the constant. Unfortunately, I don't have access to a Windows machine at the moment so am not able to investigate, but wondered if you had seen anything similar or had any ideas why it might affect this one platform?

Thanks.

NlightNFotis commented on regression/goto-analyzer/pruning-lt-nondet/test-constants.desc:
```diff
@@ -0,0 +1,8 @@
+^EXIT=0$
+^SIGNAL=0$
+\[main\.assertion\.1\] line .* assertion x < 20: SUCCESS
+
```
Hello @jezhiggins, both this and `pruning-lt-nondet-2/test-constants.desc` contain an empty line.

This looks like at least in the windows system, it fails to much as a regex. Can you delete these empty lines and see if it behaves?

I'm surprised how this is not a problem on other platforms though.

jezhiggins replied:
You're a superstar! That's got it sorted, thank you.

jezhiggins committed:
Prune TOP values in assume_eq

Found and fixed a problem in interval_abstract_value::to_constant(),
for case when interval is a valid single value, but the whole
thing is also top. This could arise in a merge of, say, [0, 0] and TOP.

martin-cs commented on regression/goto-analyzer/value-set-function-pointers-simple/main.c:
```diff
@@ -1,7 +1,6 @@
 #include <assert.h>
 
 typedef int (*fptr_t)(int);
-fptr_t fun_global, fun_global_show;
```
Would it be possible to add the reduced version of this as another test (can be another `.c` and another `.desc` in the same directory ) rather than just reducing this one?

martin-cs commented on regression/goto-analyzer/value-set-function-pointers-simple/test.desc:
```diff
@@ -1,23 +1,8 @@
-^main::1::fun2_show \(\) -> value-set-begin: ptr ->\(f\), ptr ->\(g\) :value-set-end
-^main::1::fun3_show \(\) -> value-set-begin: ptr ->\(f\), ptr ->\(g\) :value-set-end
-^fun_global_show \(\) -> value-set-begin: ptr ->\(f\), ptr ->\(g\) :value-set-end
+--variable-sensitivity --vsd-pointers value-set --show
```
Likewise here.

martin-cs commented on regression/goto-analyzer/value-set-function-pointers/test.desc:
```diff
@@ -0,0 +1,23 @@
+CORE
+main.c
+--variable-sensitivity --vsd-structs every-field --vsd-arrays every-element --vsd-pointers value-set --vsd-values set-of-constants --show --pointer-check --three-way-merge
```
Right - I see what you have done.  For the sake of a cleaner PR it might be good if the reduced one is the new one but ... it's not that big a deal.

jezhiggins replied:
Done!

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.cpp:
```diff
@@ -196,49 +225,56 @@ bool abstract_environmentt::assume(const exprt &expr, const namespacet &ns)
     }
   }
 
-  /* TODO : full implementation here
```
:tada: !!!

martin-cs commented on src/analyses/variable-sensitivity/abstract_environment.h:
```diff
@@ -96,6 +96,7 @@ class abstract_environmentt
   /// The better the implementation the more precise the results
   /// will be.
   virtual bool assume(const exprt &expr, const namespacet &ns);
+  exprt do_assume(const exprt &e, const namespacet &ns);
```
Is there any harm in this being public?  I don't think so but I thought I should check.

jezhiggins replied:
It's public so I can exercise it directly in unit tests. (Rather than via `assume` which does some other things too.) However, there's no harm can come of it been public.

martin-cs replied:
That is fine.

martin-cs commented on src/analyses/variable-sensitivity/data_dependency_context.cpp:
```diff
@@ -246,38 +246,64 @@ data_dependency_contextt::merge(abstract_object_pointert other) const
+
+  // It is critically important that we only return a newly constructed result
+  // abstract object *iff* the data has actually changed, otherwise AI may
+  // never reach a fixpoint
```
:heart:
